[playtvak] Initial support for videos from Mafra group servers

oskar456 opened the pull request:
Support for videos in articles from idnes.cz, lidovky.cz, metro.cz, as
well as video and live streams from playtvak.cz

Signed-off-by: Ondřej Caletka <ondrej@caletka.cz>

dstftw commented on youtube_dl/extractor/playtvak.py:
```diff
@@ -0,0 +1,86 @@
+        xmlinfourl = self._html_search_regex(r'Misc.videoFLV\({ data: "([^"]+)"', webpage, 'xmlinfourl')
+        xmlinfo = self._download_xml(xmlinfourl, video_id)
+
+        item = xmlinfo.find('./items/item[title]')
```
This is not supported in python 2.6. Use `utils.find_xpath_attr` instead.

dstftw replied:
This is not supported in python 2.6. Use `utils.find_xpath_attr` instead.

dstftw commented on youtube_dl/extractor/playtvak.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class PlaytvakIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:.*\.)?(playtvak|idnes|lidovky|metro)\.cz/.*\?c=(?P<id>.*)'
```
Do not capture group if you don't use it. `id` must match at least one character.

oskar456 commented:
Thanks @dstftw, hopefully fixed.

dstftw commented on youtube_dl/extractor/playtvak.py:
```diff
@@ -0,0 +1,95 @@
+
+        item = None
+        for i in xmlinfo.findall('./items/item'):
+            if i.find('title') is not None:
```
It's probably better to test for `type` tag to be equal to `video`.

oskar456 replied:
Well, it can be also a livevideo (see second test). From my observation, every video has a title.

dstftw commented on youtube_dl/extractor/playtvak.py:
```diff
@@ -0,0 +1,95 @@
+                break
+        if item is None:
+            raise ExtractorError('No suitable stream found')
+        title = xpath_text(item, './title')
```
For lives `self._live_title` should be use.

dstftw commented on youtube_dl/extractor/playtvak.py:
```diff
@@ -0,0 +1,95 @@
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        xmlinfourl = self._html_search_regex(r'Misc.videoFLV\({ data: "([^"]+)"', webpage, 'xmlinfourl')
+        xmlinfo = self._download_xml(xmlinfourl, video_id)
```
This is also available as JSON if `&type=js` is appended to the query. With JSON extra webm format is served.

oskar456 replied:
I'll look at that. What browser/OS do you use? I get XML playlist both on FIrefox and Chromium on Linux.

oskar456 commented:
Added `self._live_title` for live streams, changed the check to item type `video` or `livevideo`.

I'll look onto parsing the JS-type playlist some time later, it will probably require to drop everything and start over :)

oskar456 commented:
I have rewritten the extractor to use the JS-based playlist. There are new medium format quality streams in `webm` format and there is also HLS support for live streams, as an alternative to RTMP.

However, at least the planespotting stream (second test) uses non-standard MP3 audio, which get corrupted by hardcoded `ffmpeg` option `aac_adtstoaac`. That's why I lowered HLS stream preference.

oskar456 commented:
Fixed flake8 errors.

dstftw commented:
Broken on python 2:
```
py27yt http://www.playtvak.cz/vyzente-vosy-a-srsne-ze-zahrady-dn5-/hodinovy-manzel.aspx?c=A150730_150323_hodinovy-manzel_kuko --wri
te-pages
[Playtvak] A150730_150323_hodinovy-manzel_kuko: Downloading webpage
[Playtvak] Saving request to A150730_150323_hodinovy-manzel_kuko_http_-_www.playtvak.cz_vyzente-vosy-a-srsne-ze-zahrady-dn5-_hodinovy-manzel.aspxc=A150730_15032
3_hodinovy-manzel_kuko.dump
Traceback (most recent call last):
  File "C:\python\python2710\lib\runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "C:\python\python2710\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Dev\git\youtube-dl\master\youtube_dl\__main__.py", line 19, in <module>
    youtube_dl.main()
  File "youtube_dl\__init__.py", line 410, in main
    _real_main(argv)
  File "youtube_dl\__init__.py", line 400, in _real_main
    retcode = ydl.download(all_urls)
  File "youtube_dl\YoutubeDL.py", line 1653, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "youtube_dl\YoutubeDL.py", line 655, in extract_info
    ie_result = ie.extract(url)
  File "youtube_dl\extractor\common.py", line 287, in extract
    return self._real_extract(url)
  File "youtube_dl\extractor\playtvak.py", line 74, in _real_extract
    newquery = compat_urlparse.urlencode(qs, True)
AttributeError: 'module' object has no attribute 'urlencode'
```

oskar456 commented:
Oh s**t ;)
I fixed it using undocumented python2 urllib.unparse. Hope this works even with 2.6 (don't have such old python by hand).

dstftw commented:
I suggest you not to amend the commits but keep them separate since it's hard to track changes.

oskar456 commented:
Sorry about that. Only change was on line 75, utilizing `compat_urllib_parse`. Tests now pass on both python flavours.

dstftw commented on youtube_dl/extractor/playtvak.py:
```diff
@@ -0,0 +1,127 @@
+
+def _extract_json(code):
+    return re.sub(
+        r'(?s)^VideoPlayer.data\("", ({.*})\);?\s*?(?://[^\n]*)*$', r'\1', code)
```
What's the purpose of regex part after `;`? Can you name any example?

oskar456 replied:
Copied from [`strip_jsonp` function](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/utils.py#L1622). It should catch and drop anything after the interesting json.

dstftw commented on youtube_dl/extractor/playtvak.py:
```diff
@@ -0,0 +1,127 @@
+
+
+class PlaytvakIE(InfoExtractor):
+    _VALID_URL = r'https?://.*?(playtvak|idnes|lidovky|metro)\.cz/.*\?c=(?P<id>[A-Z][0-9]{6}_[0-9]{6}_.*)'
```
Is there any example URL for metro.cz?

oskar456 replied:
They mostly embed youtube videos, but few of them share the same publication pattern, [like this.](http://www.metro.cz/video-pod-billboardem-se-na-vltavske-roztocil-kolotoc-deti-vozil-jen-par-hodin-1hx-/metro-extra.aspx?c=A141111_173251_metro-extra_row)

dstftw commented:
Thanks, merged.
